server/src/main/java/org/opensearch/search/aggregations/bucket/terms/NumericTermsAggregator.java

```diff
@@ -255,40 +261,84 @@ private InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws
                 checkCancelled();
                 collectZeroDocEntriesIfNeeded(owningBucketOrds[ordIdx]);
                 long bucketsInOrd = bucketOrds.bucketsInOrd(owningBucketOrds[ordIdx]);
-
                 int size = (int) Math.min(bucketsInOrd, localBucketCountThresholds.getRequiredSize());
-                PriorityQueue<B> ordered = buildPriorityQueue(size);
                 B spare = null;
                 BucketOrdsEnum ordsEnum = bucketOrds.ordsEnum(owningBucketOrds[ordIdx]);
                 Supplier<B> emptyBucketBuilder = emptyBucketBuilder(owningBucketOrds[ordIdx]);
-                while (ordsEnum.next()) {
-                    long docCount = bucketDocCount(ordsEnum.ord());
-                    otherDocCounts[ordIdx] += docCount;
-                    if (docCount < localBucketCountThresholds.getMinDocCount()) {
-                        continue;
+
+                // When request size is smaller than 20% of total buckets, use priority queue to get topN buckets
+                // When request size is smaller than 20% of total buckets, use priority queue to get topN buckets
+                // partiallyBuiltBucketComparator is null for significantTerm Aggregations use case and the way buckets sorted in the
+                // priority queue is based on the significanceScore
+                if (!FeatureFlags.isEnabled(FeatureFlags.TERMS_AGGREGATION_OPTIMIZATION_ENABLE_SETTING)
```

Comment:
I think this logic needs refactoring, we need to create Enum for all strategies and provide it relevant input. It should contain logic both determine the strategy and also execute it. 
I'm thinking something like - 

[result_strat.patch](https://github.com/user-attachments/files/21555576/result_strat.patch)